failed publishing from publisher node results in orphaned lease

On a remote publisher node a publication was started and failed with this error:
```
May 15 09:52:21 biopublisher.novalocal cvmfs_server[1026994]: (bio.data.computecanada.ca) failed to publish
```
On the gateway I see it failed due to exceeding the root catalog limit which is fine, but it looks like it failed to exit cleanly ("error when quitting the receiver: worker 'quit' call failed: possible that the receiver crashed: EOF")
```
May 15 09:52:14 cvmfs-s0-bio.novalocal cvmfs_receiver[25752]: CVMFS receiver finished
May 15 09:52:14 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"info","component":"http","req_id":"cd1dc967-10d9-4b8e-9b68-ee281fd01ec2","req_dt":0.002906,"method":"POST","url":"/api/v1/leases/P62RH3I7F42HTW6K35ZYHS37CKWMR4OHSSJMANMKZULF7NL6762Q====","t0":"2023-05-15T09:52:14Z","message":"request received"}
May 15 09:52:14 cvmfs-s0-bio.novalocal cvmfs_receiver[25759]: CVMFS receiver started
May 15 09:52:14 cvmfs-s0-bio.novalocal cvmfs_receiver[25759]: CommitProcessor - lease_path: bio.data.computecanada.ca/, old hash: e33cbdcca9f58d8c19bfef07858d8df1a60ca9a7-shake128C, new hash: 94a968246e7b1f498efe0bcb44de00b1753981b1-shake128C, tag_name: generic-2023-05-15T09:52:14.063Z, tag_description:
May 15 09:52:14 cvmfs-s0-bio.novalocal cvmfs_receiver[25759]: CommitProcessor - lease_path: bio.data.computecanada.ca/, target root hash: e33cbdcca9f58d8c19bfef07858d8df1a60ca9a7-shake128
May 15 09:52:14 cvmfs-s0-bio.novalocal cvmfs_receiver[25759]: CommitProcessor - lease_path: bio.data.computecanada.ca/, merging catalogs
May 15 09:52:20 cvmfs-s0-bio.novalocal cvmfs_receiver[25759]: Runtime error during CVMFS Receiver event loop.
                                                              PANIC: /home/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/cc7/build/BUILD/cvmfs-2.10.1/cvmfs/catalog_mgr_rw.cc : 1198
                                                              catalog at / has more than 50000 entries (50974).
May 15 09:52:20 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"info","component":"actions","req_id":"cd1dc967-10d9-4b8e-9b68-ee281fd01ec2","req_dt":6126.736995,"action":"commit_lease","outcome":"runtime error: PANIC: /home/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/cc7/build/BUILD/cvmfs-2.10.1/cvmfs/catalog_mgr_rw.cc : 1198\ncatalog at / has more than 50000 entries (50974). ","action_dt":6126.665116,"message":"action complete"}
May 15 09:52:20 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"info","component":"http","req_id":"cd1dc967-10d9-4b8e-9b68-ee281fd01ec2","req_dt":6126.794572,"message":"request processed"}
May 15 09:52:20 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"error","component":"worker_pool","req_id":"cd1dc967-10d9-4b8e-9b68-ee281fd01ec2","req_dt":6128.002581,"worker_id":2,"message":"error when quitting the receiver: worker 'quit' call failed: possible that the receiver crashed: EOF"}
May 15 09:55:52 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"info","component":"http","req_id":"51f2f491-a5b3-4815-bb2f-ac4c7d2eccb3","req_dt":0.015646,"method":"DELETE","url":"/api/v1/leases/P62RH3I7F42HTW6K35ZYHS37CKWMR4OHSSJMANMKZULF7NL6762Q====","t0":"2023-05-15T09:55:52Z","message":"request received"}
May 15 09:55:52 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"info","component":"actions","req_id":"51f2f491-a5b3-4815-bb2f-ac4c7d2eccb3","req_dt":1.320291,"action":"cancel_lease","outcome":"no statistics counters for lease bio.data.computecanada.ca/","action_dt":1.140306,"message":"action complete"}
May 15 09:55:52 cvmfs-s0-bio.novalocal cvmfs_gateway[11416]: {"level":"info","component":"http","req_id":"51f2f491-a5b3-4815-bb2f-ac4c7d2eccb3","req_dt":1.385779,"message":"request processed"}
```

After that the lease remains and can not be easily cleaned up. One would have to do a manual curl call on the gateway API to cancel the lease or wait until it expires.

On the publisher node, it is not possible to do anything and abort -f does not work either:
```
$ sudo cvmfs_server transaction bio.data.computecanada.ca
Path busy. Time remaining = 10h14m59.150004416s
CernVM-FS transaction error: lease path busy
$ sudo cvmfs_server abort -f bio.data.computecanada.ca
Repository bio.data.computecanada.ca is not in a transaction
```

If a publication fails for whatever reason the gateway should handle the failure and clean up the lease.

Both components are v2.10.1

It happened again. 
```
Lease end request - error reply: {"reason":"runtime error: PANIC: /home/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/cc8/build/BUILD/cvmfs-2.10.0/cvmfs/catalog_mgr_rw.cc : 1198\ncatalog at /content/containers/qiime2-2021.11 has more than 200000 entries (226831). ","status":"error"}
SessionContext: could not commit session. Aborting.
[ERROR] Failed to commit transaction.
Statistics stored at: /var/spool/cvmfs/containers.computecanada.ca/stats.db
Synchronization failed

Executed Command:
cvmfs_swissknife sync -u /cvmfs/containers.computecanada.ca -s /var/spool/cvmfs/containers.computecanada.ca/scratch/current -c /var/spool/cvmfs/containers.computecanada.ca/rdonly -t /var/spool/cvmfs/containers.computecanada.ca/tmp -b 73a57282ee1b752d9d68c1222ae7356018e4b53a-shake128 -r gw,/srv/cvmfs/containers.computecanada.ca/data/txn,http://cvmfs-s0-soft.computecanada.ca:4929/api/v1 -w http://cvmfs-s0-soft.computecanada.ca:8000/cvmfs/containers.computecanada.ca -o /var/spool/cvmfs/containers.computecanada.ca/tmp/manifest -e shake128 -Z default -C /etc/cvmfs/repositories.d/containers.computecanada.ca/trusted_certs -N containers.computecanada.ca -K /etc/cvmfs/keys/containers.computecanada.ca.pub -L -D generic-2023-07-24T16:11:48Z -H /etc/cvmfs/keys/containers.computecanada.ca.gw -P /var/spool/cvmfs/containers.computecanada.ca/session_token -f overlayfs -p -l 4194304 -a 8388608 -h 16777216 -i
[libuser-containers@publisher-node ~]$ timed out waiting for input: auto-logout
Error from gateway: 'no statistics counters for lease containers.computecanada.ca/content'
force abort, continue despite error while trying to drop lease, removing session token. Error: gateway doesn't recognize the lease or cannot drop it
```

--- a/test/src/817-gateway_abort/main
+++ b/test/src/817-gateway_abort/main
@@ -1,0 +1,26 @@
+cvmfs_test_name="Lease cleanup when the publishing process aborts on the gateway"
+cvmfs_test_autofs_on_startup=false
+cvmfs_test_suites="quick"
+
+
+cvmfs_run_test() {
+    set_up_repository_gateway || return 1
+
+    echo "*** Starting initial transaction"
+    cvmfs_server transaction test.repo.org || return 10
+    cvmfs_server publish test.repo.org     || return 11
+
+    echo "*** Let transaction crash on the gateway side"
+    cvmfs_server transaction test.repo.org || return 20
+    sudo mv /etc/cvmfs/keys/test.repo.org.key /etc/cvmfs/keys/test.repo.org.key.hidden
+    cvmfs_server publish test.repo.org && return 21
+
+    echo "*** Try to recover"
+    cvmfs_server abort -f test.repo.org || return 30
+    sudo mv /etc/cvmfs/keys/test.repo.org.key.hidden /etc/cvmfs/keys/test.repo.org.key
+    cvmfs_server transaction test.repo.org || return 31
+    cvmfs_server publish test.repo.org || return 32
+
+    return 0
+}
+

